lib/tether/toolbox.go

```diff
@@ -237,3 +270,161 @@ func (t *Toolbox) halt() error {
+	skipdata, _ := strconv.ParseBool(u.Query().Get(vsphere.SkipDataQueryName))
+
+	if diskLabel != "" && filterSpec != "" {
+		op := trace.NewOperation(context.Background(), "ToolboxOnlineDataSource: %s", u.String())
+		op.Debugf("Writing to archive from %s: %s", u.Path, u.String())
+
+		spec, err := archive.DecodeFilterSpec(op, &filterSpec)
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+
+		// get the container fs mount
+		diskPath, err := mountDiskLabel(diskLabel)
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+		defer unmount(op, diskPath)
+
+		var rc io.ReadCloser
+		if skiprecurse {
+			// we only want a single file - this is a hack while we're abusing Diff, but
+			// accomplish this by generating a single entry ChangeSet
+			changes := []dar.Change{
+				{
+					Kind: dar.ChangeModify,
+					Path: u.Path,
+				},
+			}
+
+			rc, err = archive.Tar(op, diskPath, changes, spec, !skipdata, false)
+		} else {
+			rc, err = archive.Diff(op, diskPath, "", spec, !skipdata, false)
+		}
+
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer rc.Close()
+			defer wg.Done()
+			tr := tar.NewReader(rc)
+			for {
+				hdr, err := tr.Next()
+				if err == io.EOF {
+					err = nil
+					op.Debugf("Finished writing to archive from %s: %s with error %#v", u.Path, u.String(), err)
+					break
+				}
+				if err != nil {
+					op.Debugf("error writing tar: %s", err.Error())
+					break
+				}
+				op.Debugf("Writing header: %#s", *hdr)
+				err = tw.WriteHeader(hdr)
+				if err != nil {
+					op.Debugf("error writing tar header: %s", err.Error())
+					break
+				}
+				_, err = io.Copy(tw, tr)
+				if err != nil {
+					op.Debugf("error writing tar contents: %s", err.Error())
+					break
+				}
+			}
+		}()
+		wg.Wait()
+		return err
+	}
+	return defaultArchiveHandler.Write(u, tw)
+}
+
+func mountDiskLabel(label string) (string, error) {
+	// We know the vmdk will always be attached at '/'
+	if label == "containerfs" {
+		return "/", nil
+	}
+
+	// otherwise, label represents a volume that needs to be mounted
+	path := path.Join("/dev/disk/by-label/", label)
+	_, err := os.Lstat(path)
+	// label does not exist
+	if err != nil {
+		return "", err
+	}
+
+	// label exists, mount the device to a tmp directory
+	tmpDir, err := ioutil.TempDir("", fmt.Sprintf("toolbox-%s", label))
+	if err := Sys.Syscall.Mount(path, tmpDir, ext4FileSystemType, syscall.MS_NOATIME, ""); err != nil {
+		return "", fmt.Errorf("faild to mount %s to %s: %s", path, tmpDir, err)
+	}
+
+	return tmpDir, nil
+}
+
+func unmount(op trace.Operation, unmountPath string) {
+	// don't unmount the root vmdk
+	if unmountPath == "/" {
+		return
+	}
+
+	// unmount the disk from the temporary directory
+	if err := Sys.Syscall.Unmount(unmountPath, syscall.MNT_DETACH); err != nil {
+		op.Debugf("failed to unmount %s: %s", unmountPath, err.Error())
+	}
```

Comment:
This was a problem in the appliance also - I replaced the recursive delete with a directory delete. I see no circumstance under which we actually _want_ recursive deletion here.